File: src/main/java/com/example/librarymanagement/service/impl/ReservationServiceImpl.java
package com.example.librarymanagement.service.impl;

import com.example.librarymanagement.exception.EntityNotFoundException;
import com.example.librarymanagement.model.entity.Book;
import com.example.librarymanagement.model.entity.BookPresence;
import com.example.librarymanagement.model.entity.Library;
import com.example.librarymanagement.model.entity.Reservation;
import com.example.librarymanagement.model.entity.User;
import com.example.librarymanagement.model.enums.Availability;
import com.example.librarymanagement.repository.BookRepository;
import com.example.librarymanagement.repository.ReservationRepository;
import com.example.librarymanagement.service.BookPresenceService;
import com.example.librarymanagement.service.ReservationService;
import jakarta.transaction.Transactional;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
@RequiredArgsConstructor
public class ReservationServiceImpl implements ReservationService {
    private final ReservationRepository reservationRepository;
    private final BookRepository bookRepository;
    private final BookPresenceService bookPresenceService;

    @Override
    public List<Reservation> getReservationsByLibraryId(Long libraryId) {
        return reservationRepository.findAllByLibraryId(libraryId);
    }

    @Override
    public List<Reservation> getReservationsByUserId(Long userId) {
        return reservationRepository.findAllByUserId(userId);
    }

    @Override
    public List<Reservation> getReservationsByBookIdAndUser(Long bookId, User user) {
        return reservationRepository.findAllByBookIdAndUserId(bookId, user.getId());
    }

    @Override
    @Transactional
    public List<Reservation> doReservationBook(User user, Long libraryId, Long bookId) {
        Library library;
        List<BookPresence> bookPresenceList;
        Book book = bookRepository.findById(bookId).orElseThrow(() -> new EntityNotFoundException("Book"));

        if (libraryId != null) {
            bookPresenceList = bookPresenceService.getAllBookByLibraryIdAndBookId(libraryId, bookId);
            library = bookPresenceList.get(0).getLibrary();
        } else {
            bookPresenceList = bookPresenceService.getByBookId(bookId);
            library = null;
        }

        Optional<BookPresence> bookPresence = bookPresenceList.stream()
                .filter(i -> i.getAvailability().equals(Availability.AVAILABLE)).findFirst();

        if (bookPresence.isPresent()) {
            bookPresenceService.addUserToBook(user, libraryId, bookId);
        } else {
            Reservation reservation = new Reservation();
            reservation.setUser(user);
            reservation.setBook(book);
            reservation.setLibrary(library);

            book.getReservations().add(reservation);
            user.getReservations().add(reservation);

            reservationRepository.save(reservation);
        }
        return user.getReservations();
    }

    @Override
    @Transactional
    public void removeReservation(User user, Long bookId) {
        List<Reservation> reservations = getReservationsByBookIdAndUser(bookId, user);
        for (Reservation reservation : reservations) {
            reservation.getBook().getReservations().remove(reservation);
            reservation.getUser().getReservations().remove(reservation);
            deleteReservationById(reservation.getId());
        }
    }

    @Override
    public void deleteReservationById(Long id) {
        reservationRepository.findById(id)
                .ifPresent(reservation -> {
                    reservation.setBook(null);
                    reservation.setUser(null);
                    reservation.setLibrary(null);

                    reservationRepository.delete(reservation);
                });
    }
}

File: src/main/java/com/example/librarymanagement/service/impl/ReservationServiceImpl.kt
package com.example.librarymanagement.service.impl

import com.example.librarymanagement.exception.EntityNotFoundException
import com.example.librarymanagement.exception.ExistingReservationException
import com.example.librarymanagement.model.domain.Reservation
import com.example.librarymanagement.model.domain.ReservationOutcome
import com.example.librarymanagement.model.enums.Availability
import com.example.librarymanagement.repository.ReservationRepository
import com.example.librarymanagement.service.BookPresenceService
import com.example.librarymanagement.service.JournalService
import com.example.librarymanagement.service.LibraryService
import com.example.librarymanagement.service.ReservationService
import jakarta.transaction.Transactional
import org.springframework.stereotype.Service

@Service
class ReservationServiceImpl(
    private val reservationRepository: ReservationRepository,
    private val bookPresenceService: BookPresenceService,
    private val libraryService: LibraryService,
    private val journalService: JournalService
) : ReservationService {

    override fun getReservationsByUserId(userId: String): List<Reservation> =
        reservationRepository.findAllByUserId(userId)

    @Transactional
    override fun reserveBook(userId: String, libraryId: String?, bookId: String): ReservationOutcome {
        if (reservationRepository.existsByBookIdAndUserId(bookId, userId)) {
            throw ExistingReservationException(bookId, userId)
        }

        val libraryForReservation = determineLibraryForReservation(libraryId, bookId)

        val availableBookPresence = bookPresenceService
            .getAllBookPresencesByLibraryIdAndBookId(libraryForReservation, bookId)
            .firstOrNull { it.availability == Availability.AVAILABLE }

        return availableBookPresence?.let {
            bookPresenceService.addUserToBook(userId, availableBookPresence.libraryId, bookId)
            ReservationOutcome.Journals(journalService.getJournalByUserId(userId))
        } ?: run {
            reservationRepository.save(Reservation(userId = userId, libraryId = libraryForReservation, bookId = bookId))
            ReservationOutcome.Reservations(getReservationsByUserId(userId))
        }
    }

    @Transactional
    override fun cancelReservation(userId: String, bookId: String) {
        reservationRepository.findAllByBookIdAndUserId(bookId, userId)
            .map { it.id?.let { it1 -> reservationRepository.deleteById(it1) } }
    }

    override fun deleteReservationById(id: String) {
        reservationRepository.findById(id)?.let { reservationRepository.deleteById(id) }
    }

    private fun determineLibraryForReservation(libraryId: String?, bookId: String): String {
        return libraryId?.let {
            libraryService.getLibraryById(libraryId)
            if (!bookPresenceService.existsBookPresenceByBookIdAndLibraryId(bookId, libraryId)) {
                throw EntityNotFoundException("Presence of book")
            }
            libraryId
        } ?: run {
            findLibraryWithFewestReservations(bookId) ?: throw EntityNotFoundException("Libraries with book")
        }
    }

    private fun findLibraryWithFewestReservations(bookId: String): String? {
        return bookPresenceService.getAllByBookId(bookId)
            .map { bookPresence ->
                val reservationCount = reservationRepository.findAllByBookIdAndLibraryId(
                    bookPresence.bookId,
                    bookPresence.libraryId
                ).size
                bookPresence.libraryId to reservationCount
            }
            .minByOrNull { it.second }
            ?.first
    }
}
